### ISSUE TITLE
Material uniqueness and identification

### ISSUE THREAD

**Issue by bam241** [pyne/pyne#1182]
Trying to implement the MaterialLibrary in the C++ API this concern has been raised (see #1104 )

>>Material uniqueness and identification
>How this is decided is unclear to me. What unique properties are materials required to have? Right now we require unique material IDs in the metadata (it seems), but their names are how we find and remove them from a material library - so which of those two pieces of information uniquely identifies a PyNE material? We currently hash a material using its address in Python memory, which seems like it should be enough? Whether two materials are equivalent or not could depend on some combination of their composition, metadata, etc. in a different way, perhaps? But for looking up a material in a material library for removal/containment, the hash of the object seems to make the most sense to me.

>The unique identifier I am using (and that was somehow use in the python interface was the material name) as it was automatically set (if lacking) when writing the MaterialLibrary in hdf5.

>we can certainly use something else, I really like the hash reference.
(Before I start implementing it, @makeclean @gonuke any thought about this ?)

The bottom question is:
Shall we keep name as unique identifier for material in the MaterialLibrary ?
Or shall we rely on the Material object hash as an unique identifier for the Material in the MaterialLibrary ?

I personnel like the Hash idea, but it rises some question, how to query a material in the MaterialLibrary, by name ? What if there is multiple of them sharing the same name ?

**Comment by pshriwise** [comment]
Agreed on using the `id()` of the object to uniquely identify a material object, @bam241. Maybe even better would be to use the address of the underlying C++ object. Correct me if I'm wrong, but if you have two Python `Material` objects with the same `inst_` attribute. They are going to behave exactly the same, right? But maybe material creation and copying is setup such that this can never happen and it isn't an issue.

Material equivalence is a little trickier.... do we do this by composition only? Include density? All atributes? I don't have an answer in mind for that one.

**Comment by pshriwise** [comment]
As far as querying a material by name in the library goes, a convenience method like `get_mat_by_name` (and maybe another like `get_mat_by_id` also) on the `MaterialLibrary` would be useful?

**Comment by gonuke** [comment]
I don’t think we should compare materials by their composition.  A user may have reasons to have two identical materials with different names/ids in their library.

I think name in the dictionary/map should be the UID, and metadata should be ignored/optional.

Is there a use case that I’m missing?

**Comment by gonuke** [comment]
I suppose the use case is #1183

**Comment by pshriwise** [comment]
> I don’t think we should compare materials by their composition. A user may have reasons to have two identical materials with different names/ids in their library

Agreed. I don't think what I'm suggesting would keep this from happening. A user could still have two different `Material` objects that will behave the same in simulation. When referencing the composition, I'm trying to figure out the set of information that will allow us to determine if they'll behave the same.

```python
m1 = Material(<composition>)
m2 = copy(m1)

m1 is m1 # True - same object
m1 == m1 # True - will behave the same in sim

m1 is m2 # False - not the same object
m1 == m2 # True - will behave the same in sim
```
This might be simulation/physics code dependent and depend on metadata which, I agree, should be optional. Which leaves mainly the composition to compare... maybe I'm off track there. IDK.

> I think name in the dictionary/map should be the UID, and metadata should be ignored/optional.

Can you elaborate on this a bit? Should name be key?

**Comment by gonuke** [comment]
Regarding comparisons: I was only thinking in the context of the `MaterialLibrary` class and how we might use these notions.  

I agree that it's an interesting question from the point of view of the `Material` class and whether/how we should define operators like this.  Since it may be application specific, we might need to define application-specific versions of these.

Regarding UID/key: I think the key for a `Material` object in a `MaterialLibrary`  dictionary/map should be it's name.

**Comment by pshriwise** [comment]
Comparisons:
I see! I think I misunderstood your concern. Yes. From the standpoint of having a collection of objects, we should compare the objects themselves.

And yeah, I was wondering about application-specific comparisons too. I'd like to think that there's some way to know that if one were to prepare two materials for a specific application in the same way (via metadata updates, density setting, etc.) that they would behave the same. Maybe this would be what our `==` could represent and that could exist in conjunction with application-specific comparision methods?

UID/Key: 
Same page!

**Comment by gonuke** [comment]
@pshriwise, @bam241 : I don't think this is resolved, but not sure what the action is, either....?
